File: knight_tour.cpp
#include <iostream>

using namespace std;
struct chessSquare {
    int row = 0;
    int column = 0;
    int mark = 0;
};

template<class t>
class Stack {
    t *arr;
    int top;
public:
    int size;

    Stack(int s = 100) {
        arr = new t[s];
        top = -1;
        size = 0;
    };

    bool isEmpty() {
        return top < 0;
    }

    bool isFull() {
        return top >= 99;
    }

    void push(t value) {
        if (!isFull()) {
            top++;
            arr[top] = value;
            size++;
        }
    }

    t pop() {
        if (!isEmpty()) {
            t temp = arr[top];
            size--;
            top--;
            return temp;
        }
    }

    t Top() {
        if (!isEmpty())
            return arr[top];
    }

};

class knightTour {
    chessSquare chess[9][9];
    int colMove[8] = {1, 1, -1, -1, 2, 2, -2, -2};
    int rowMove[8] = {2, -2, -2, 2, 1, -1, -1, 1};
    Stack<chessSquare> tour;
public:
    knightTour(int startRow, int startColumn) {

        chess[startRow][startColumn].row = startRow;
        chess[startRow][startColumn].column = startColumn;
        chess[startRow][startColumn].mark = 1;
        tour.push(chess[startRow][startColumn]);
    }

    int Tour() {

        // Base
        if (isFinished())
            return 1;

        if (tour.isEmpty())
            return 0;

        chessSquare current = tour.Top();

        for (int i = 0; i < 8; i++) {
            chessSquare nextMove = next(current, i);

            if (nextMove.mark != -1) {
                nextMove.mark = 1;
                tour.push(nextMove);
                int a = Tour();
                if (a == 1)
                    return 1;
            }
        }

        current.mark = 0;
        tour.pop();
        return 0;
    }

    chessSquare next(chessSquare current, int i) {
        chessSquare tempo;
        tempo.mark = -1;
        chessSquare *temp = nullptr;
        int tempRow = current.row + rowMove[i];
        int tempCol = current.column + colMove[i];

        if (tempRow >= 0 && tempRow <= 7 && tempCol >= 0 && tempCol <= 7) {
            temp = &chess[tempRow][tempCol];
            temp->row = tempRow;
            temp->column = tempCol;
            if (temp->mark != 1) {
                return *temp;
            }
        }

        return tempo;
    }

    bool isFinished() {
        return tour.size == 64;
    }

    Stack<chessSquare> getTour() {
        return tour;
    }

};

int main() {
    string p;
    int startRow, startColumn;
    cout << "input type :  ( row , col )" << endl;
    cin >> p >> startRow >> p >> startColumn >> p;
    knightTour obj(startRow, startColumn);
    obj.Tour();
    Stack<chessSquare> tmp = obj.getTour();
    Stack<chessSquare> path;
    while (!tmp.isEmpty()) {
        path.push(tmp.pop());
    }
    for (int i = 0; i < 64; ++i) {
        chessSquare temp = path.pop();
        cout << i + 1 << " - " << "( " << temp.row << " , " << temp.column << " )" << endl;
    }
    return 0;
}

File: knight_tour.py
class knight_tour:
    def __init__(self, row_number, col_number):
        self.row_number = row_number
        self.col_number = col_number
        self.row_move = [1, 1, -1, -1, 2, 2, -2, -2]
        self.col_move = [2, -2, -2, 2, 1, -1, -1, 1]
        self.chess = []
        for i in range(1, row_number + 1):
            for j in range(1, col_number + 1):
                new_cell = {"row": i, "col": j, "mark": None}
                self.chess.append(new_cell)
        self.tour = []

    def is_finished(self):
        return (
            self.tour[-1]["row"] == self.last_cell["row"]
            and self.tour[-1]["col"] == self.last_cell["col"]
        )

    def mark(self, marked_cell, mark):
        for cell in self.chess:
            if cell["row"] == marked_cell["row"] and cell["col"] == marked_cell["col"]:
                cell["mark"] = mark

    def next(self, current, i):
        row = current["row"] + self.row_move[i]
        col = current["col"] + self.col_move[i]
        for cell in self.chess:
            if cell["row"] == row and cell["col"] == col:
                if cell["mark"] is None:
                    cell["mark"] = True
                    new_cell = {"row": row, "col": col}
                    return new_cell
        return False

    def run_tour(self):
        if self.is_finished():
            return True

        current = self.tour[-1]

        for i in range(8):
            next_cell = self.next(current, i)
            if next_cell:
                self.tour.append(next_cell)
                if self.run_tour():
                    return True
                else:
                    self.tour.pop()
                    self.mark(next_cell, False)
        self.mark(current, False)
        return False

    def start_tour(self, first_cell, last_cell):
        first_cell["mark"] = True
        last_cell["mark"] = True
        self.last_cell = last_cell
        self.tour.append(first_cell)
        self.mark(first_cell, True)
        if self.run_tour():
            return self.tour
        else:
            return False


def print_chess(chess, row, col):
    for i in range(row):
        for j in range(col):
            print(chess[0]["mark"], end=" ")
            chess.remove(chess[0])
        print()


dims = input().split(" ")
row_number = int(dims[0])
col_number = int(dims[1])

start = input().split(" ")
start_row = int(start[0])
start_col = int(start[1])
start_cell = {"row": start_row, "col": start_col, "mark": True}

stop = input().split(" ")
stop_row = int(stop[0])
stop_col = int(stop[1])
stop_cell = {"row": stop_row, "col": stop_col, "mark": True}

knight_tour = knight_tour(row_number, col_number)

tour = knight_tour.start_tour(start_cell, stop_cell)
if tour:
    print_chess(knight_tour.chess, row_number, col_number)
else:
    print("No Way!")

